README.md
# TCPDF
*PHP PDF Library*

[![Donate via PayPal](https://img.shields.io/badge/donate-paypal-87ceeb.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_donations&currency_code=GBP&business=paypal@tecnick.com&item_name=donation%20for%20TCPDF%20project)
*Please consider supporting this project by making a donation via [PayPal](https://www.paypal.com/cgi-bin/webscr?cmd=_donations&currency_code=GBP&business=paypal@tecnick.com&item_name=donation%20for%20TCPDF%20project)*

* **category**    Library
* **author**      Nicola Asuni <info@tecnick.com>
* **copyright**   2002-2016 Nicola Asuni - Tecnick.com LTD
* **license**     http://www.gnu.org/copyleft/lesser.html GNU-LGPL v3 (see LICENSE.TXT)
* **link**        http://www.tcpdf.org
* **source**      https://github.com/tecnickcom/TCPDF


## IMPORTANT
A new version of this library is under development at https://github.com/tecnickcom/tc-lib-pdf and as a consequence this version will not receive any additional development or support.
This version should be considered obsolete, new projects should use the new version as soon it will become stable.



## Description

PHP library for generating PDF documents on-the-fly.

### Main Features:
* no external libraries are required for the basic functions;
* all standard page formats, custom page formats, custom margins and units of measure;
* UTF-8 Unicode and Right-To-Left languages;
* TrueTypeUnicode, OpenTypeUnicode v1, TrueType, OpenType v1, Type1 and CID-0 fonts;
* font subsetting;
* methods to publish some XHTML + CSS code, Javascript and Forms;
* images, graphic (geometric figures) and transformation methods;
* supports JPEG, PNG and SVG images natively, all images supported by GD (GD, GD2, GD2PART, GIF, JPEG, PNG, BMP, XBM, XPM) and all images supported via ImagMagick (http://www.imagemagick.org/script/formats.php)
* 1D and 2D barcodes: CODE 39, ANSI MH10.8M-1983, USD-3, 3 of 9, CODE 93, USS-93, Standard 2 of 5, Interleaved 2 of 5, CODE 128 A/B/C, 2 and 5 Digits UPC-Based Extension, EAN 8, EAN 13, UPC-A, UPC-E, MSI, POSTNET, PLANET, RMS4CC (Royal Mail 4-state Customer Code), CBC (Customer Bar Code), KIX (Klant index - Customer index), Intelligent Mail Barcode, Onecode, USPS-B-3200, CODABAR, CODE 11, PHARMACODE, PHARMACODE TWO-TRACKS, Datamatrix, QR-Code, PDF417;
* JPEG and PNG ICC profiles, Grayscale, RGB, CMYK, Spot Colors and Transparencies;
* automatic page header and footer management;
* document encryption up to 256 bit and digital signature certifications;
* transactions to UNDO commands;
* PDF annotations, including links, text and file attachments;
* text rendering modes (fill, stroke and clipping);
* multiple columns mode;
* no-write page regions;
* bookmarks, named destinations and table of content;
* text hyphenation;
* text stretching and spacing (tracking);
* automatic page break, line break and text alignments including justification;
* automatic page numbering and page groups;
* move and delete pages;
* page compression (requires php-zlib extension);
* XOBject Templates;
* Layers and object visibility.
* PDF/A-1b support.

### Third party fonts:
This library may include third party font files released with different licenses.

All the PHP files on the fonts directory are subject to the general TCPDF license (GNU-LGPLv3),
they do not contain any binary data but just a description of the general properties of a particular font.
These files can be also generated on the fly using the font utilities and TCPDF methods.

All the original binary TTF font files have been renamed for compatibility with TCPDF and compressed using the gzcompress PHP function that uses the ZLIB data format (.z files).

The binary files (.z) that begins with the prefix "dejavu" have been extracted from the DejaVu fonts 2.33 (Bitstream) collection.
For the details of Copyright, License and other information, please check the files inside the directory fonts/dejavu-fonts-ttf-2.33
Link : http://dejavu-fonts.org

### ICC profile:

TCPDF includes the sRGB.icc profile from the icc-profiles-free Debian package:
https://packages.debian.org/source/stable/icc-profiles-free


## Developer(s) Contact

* Nicola Asuni <info@tecnick.com>


CHANGELOG.TXT
Unreleased
Refactored classes and project structure to use namespaces
Removed K_TCPDF_THROW_EXCEPTION_ERROR - exceptions are always thrown
Removed TCPDFBarcode::getBarcodeSVG(), TCPDFBarcode::getBarcodePNG()
Removed TCPDF2DBarcode::getBarcodePNG() TCPDF2DBarcode::getBarcodeSVG()

6.2.12 (2015-09-12)
	- fix composer package name to tecnickcom/tcpdf

6.2.11 (2015-08-02)
	- Bug #1070 "PNG regression in 6.2.9 (they appear as their alpha channel)" was fixed.
	- Bug #1069 "Encoded SRC URLs in <img> tags don't work anymore" was fixed.

6.2.10 (2015-07-28)
	- Minor mod to PNG parsing.
	- Make dependency on mcrypt optional.

6.2.8 (2015-04-29)
	- Removed unwanted file.

6.2.7 (2015-04-28)
	- Merged PR 17: Avoid warning when iterating a non-array variable.
	- Merged PR 16: Improve MuliCell param definition.
	- Improved column check (PR 15).
	- Merged PR 11: Use stream_is_local instead of limit to file://.
	- Merged PR 10: ImageMagick link on README.txt.

6.2.6 (2015-01-28)
	- Bug #1008 "UTC offset sing breaks PDF/A-1b compliance" was fixed.

6.2.5 (2015-01-24)
	- Bug #1019 "$this in static context" was fixed.
	- Bug #1015 "Infinite loop in getIndirectObject method of parser" was fixed.

6.2.4 (2015-01-08)
	- fix warning related to empty K_PATH_URL.
	- fix error when a $table_colwidths key is not set.

6.2.3 (2014-12-18)
	- New comment.
	- Moved the K_PATH_IMAGES definition in tcpdf_autoconfig.

6.2.2 (2014-12-18)
	- Fixed mispelled words.
	- Fixed version number.

6.2.1 (2014-12-18)
	- The constant K_TCPDF_THROW_EXCEPTION_ERROR is now set to false in the default configuration file.
	- An issue with the _destroy() method was fixed.

6.2.0 (2014-12-10)
	- Bug #1005 "Security Report, LFI posting internal files externally abusing default parameter" was fixed.
	- Static methods serializeTCPDFtagParameters() and unserializeTCPDFtagParameters() were moved as non static to the main TCPDF class (see changes in example n. 49).
	- Deprecated methods were removed, please use the equivalents defined in other classes (i.e. TCPDF_STATIC and TCPDF_FONTS).
	- The constant K_TCPDF_CALLS_IN_HTML is now set by default to FALSE.
	- DLE, DLX and DLP page format was added.
	- Page format are now defined as a public property in TCPDF_STATIC.

6.1.1 (2014-12-09)
	- Fixed bug with the register_shutdown_function().

6.1.0 (2014-12-07)
	- The method TCPDF_STATIC::getRandomSeed() was improved.
	- The disk caching feature was removed.
	- Bug #1003 "Backslashes become duplicated in table, using WriteHTML" was fixed.
	- Bug #1002 "SVG radialGradient within non-square Rect" was fixed.

6.0.099 (2014-11-15)
	- Added basic support for nested SVG images (adapted PR from SamMousa).
	- A bug related to setGDImageTransparency() was fixed (thanks to Maarten Boerema).

6.0.098 (2014-11-08)
	- Bug item #996 "getCharBBox($char) returns incorrect results for TTF glyphs without outlines" was fixed.
	- Bug item #991 "Text problem with SVG" was fixed (only the font style part).

6.0.097 (2014-10-20)
	- Bug item #988 "hyphenateText - charmin parameter not work" was fixed.
	- New 1D barcode method to print pre-formatted IMB - Intelligent Mail Barcode - Onecode - USPS-B-3200.

6.0.096 (2014-10-06)
	- Bug item #982 "Display style is not inherited in SVG" was fixed.
	- Bug item #984 "Double quote url in CSS" was fixed.

6.0.095 (2014-10-02)
	- Bug item #979 "New Timezone option overwriting current timezone" was fixed.

6.0.094 (2014-09-30)
	- Bug item #978 "Variable Undefined: $cborder" was fixed.

6.0.093 (2014-09-02)
	- Security fix: some serialize/unserialize methods were replaced with json_encode/json_decode to avoid a potential object injection with user supplied content. Thanks to ownCloud Inc. for reporting this issue.
	- K_TIMEZONE constant was added to the default configuration to suppress date-time warnings.

6.0.092 (2014-09-01)
	- Bug item #956 "Monospaced fonts are not alignd at the baseline" was fixed.
	- Bug item #964 "Problem when changing font size" was fixed.
	- Bug item #969 "ImageSVG with radialGradient problem" was fixed.
	- sRGB.icc file was replaced with the one from the Debian package icc-profiles-free (2.0.1+dfsg-1) 

6.0.091 (2014-08-13)
	- Issue #325"Division by zero when css fontsize equals 0" was fixed.

6.0.090 (2014-08-08)
	- Starting from this version TCPDF is also available in GitHub at https://github.com/tecnickcom/TCPDF
	- Function getmypid() was removed for better compatibility with shared hosting environments.
	- Support for pulling SVG stroke opacity value from RGBa color was mergeg [adf006].
	- Bug item #951 "HTML Table within TCPDF columns doesnt flow correctly on page break ..." was fixed. 

6.0.089 (2014-07-16)
	- Bug item #948 "bottom line of rowspan cell not work correctly" was fixed.

6.0.088 (2014-07-09)
	- Bug item #946 "Case sensitive type check causes broken match for SVG" was fixed.
	- Bug item #945 "Imagick load doesn't account for passed data string " was fixed.

6.0.087 (2014-06-25)
	- A bug affecting fitcell option in Multicell was fixed.

6.0.086 (2014-06-20)
	- Bug item #938 "Hyphenation-dash extends outside of cell" was fixed (collateral effect).

6.0.085 (2014-06-19)
	- Some example images were replaced.
	- A race condition bug was fixed.
	- Bug item #938 "Hyphenation-dash extends outside of cell" was fixed.

6.0.084 (2014-06-13)
	- A bug related to MultiCell fitcell feature was fixed.
	- Bug item #931 "Documentation error for setPageFormat()" was fixed.

6.0.083 (2014-05-29)
	- Bug item #928 "setHtmlVSpace with HR element" was fixed.

6.0.082 (2014-05-23)
	- Bug item #926 "test statement instead of assignment used in tcpdf_fonts.php" was fixed.
	- Bug item #925 "924 transparent images bug" was fixed.

6.0.081 (2014-05-22)
	- Bug item #922 "writehtml tables thead repeating" was fixed.
	- Patch #71 "External and internal links, local and remote" wa applied.

6.0.080 (2014-05-20)
	- Bug item #921 "Fatal error in hyphenateText() function" was fixed.

src/IMAGES.php
<?php
//============================================================+
// File name   : tcpdf_images.php
// Version     : 1.0.005
// Begin       : 2002-08-03
// Last Update : 2014-11-15
// Author      : Nicola Asuni - Tecnick.com LTD - www.tecnick.com - info@tecnick.com
// License     : GNU-LGPL v3 (http://www.gnu.org/copyleft/lesser.html)
// -------------------------------------------------------------------
// Copyright (C) 2002-2014 Nicola Asuni - Tecnick.com LTD
//
// This file is part of TCPDF software library.
//
// TCPDF is free software: you can redistribute it and/or modify it
// under the terms of the GNU Lesser General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// TCPDF is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
// See the GNU Lesser General Public License for more details.
//
// You should have received a copy of the License
// along with TCPDF. If not, see
// <http://www.tecnick.com/pagefiles/tcpdf/LICENSE.TXT>.
//
// See LICENSE.TXT file for more information.
// -------------------------------------------------------------------
//
// Description :
//   Static image methods used by the TCPDF class.
//
//============================================================+
namespace TCPDF;

/**
 * @file
 * This is a PHP class that contains static image methods for the TCPDF class.<br>
 * @package com.tecnick.tcpdf
 * @author Nicola Asuni
 * @version 1.0.005
 */

/**
 * @class TCPDF\IMAGES
 * Static image methods used by the TCPDF class.
 * @package com.tecnick.tcpdf
 * @brief PHP class for generating PDF documents without requiring external extensions.
 * @version 1.0.005
 * @author Nicola Asuni - info@tecnick.com
 */
class IMAGES
{

    /**
     * Array of hinheritable SVG properties.
     * @since 5.0.000 (2010-05-02)
     * @public static
     */
    public static $svginheritprop = [
        'clip-rule',
        'color',
        'color-interpolation',
        'color-interpolation-filters',
        'color-profile',
        'color-rendering',
        'cursor',
        'direction',
        'display',
        'fill',
        'fill-opacity',
        'fill-rule',
        'font',
        'font-family',
        'font-size',
        'font-size-adjust',
        'font-stretch',
        'font-style',
        'font-variant',
        'font-weight',
        'glyph-orientation-horizontal',
        'glyph-orientation-vertical',
        'image-rendering',
        'kerning',
        'letter-spacing',
        'marker',
        'marker-end',
        'marker-mid',
        'marker-start',
        'pointer-events',
        'shape-rendering',
        'stroke',
        'stroke-dasharray',
        'stroke-dashoffset',
        'stroke-linecap',
        'stroke-linejoin',
        'stroke-miterlimit',
        'stroke-opacity',
        'stroke-width',
        'text-anchor',
        'text-rendering',
        'visibility',
        'word-spacing',
        'writing-mode'
    ];

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

    /**
     * Return the image type given the file name or array returned by getimagesize() function.
     * @param $imgfile (string) image file name
     * @param $iminfo (array) array of image information returned by getimagesize() function.
     * @return string image type
     * @since 4.8.017 (2009-11-27)
     * @public static
     */
    public static function getImageFileType($imgfile, $iminfo = [])
    {
        $type = '';
        if (isset($iminfo['mime']) and !empty($iminfo['mime'])) {
            $mime = explode('/', $iminfo['mime']);
            if ((count($mime) > 1) and ($mime[0] === 'image') and (!empty($mime[1]))) {
                $type = strtolower(trim($mime[1]));
            }
        }
        if (empty($type)) {
            $fileinfo = pathinfo($imgfile);
            if (isset($fileinfo['extension']) and (!TCPDFSTATIC::empty_string($fileinfo['extension']))) {
                $type = strtolower(trim($fileinfo['extension']));
            }
        }
        if ($type === 'jpg') {
            $type = 'jpeg';
        }
        return $type;
    }

    /**
     * Set the transparency for the given GD image.
     * @param $new_image (image) GD image object
     * @param $image (image) GD image object.
     * return GD image object.
     * @since 4.9.016 (2010-04-20)
     * @public static
     */
    public static function setGDImageTransparency($new_image, $image)
    {
        // default transparency color (white)
        $tcol = ['red' => 255, 'green' => 255, 'blue' => 255];
        // transparency index
        $tid = imagecolortransparent($image);
        $palletsize = imagecolorstotal($image);
        if (($tid >= 0) and ($tid < $palletsize)) {
            // get the colors for the transparency index
            $tcol = imagecolorsforindex($image, $tid);
        }
        $tid = imagecolorallocate($new_image, $tcol['red'], $tcol['green'], $tcol['blue']);
        imagefill($new_image, 0, 0, $tid);
        imagecolortransparent($new_image, $tid);
        return $new_image;
    }

    /**
     * Convert the loaded image to a PNG and then return a structure for the PDF creator.
     * This function requires GD library and write access to the directory defined on K_PATH_CACHE constant.
     * @param $image (image) Image object.
     * @param $tempfile (string) Temporary file name.
     * return image PNG image object.
     * @since 4.9.016 (2010-04-20)
     * @public static
     */
    public static function _toPNG($image, $tempfile)
    {
        // turn off interlaced mode
        imageinterlace($image, 0);
        // create temporary PNG image
        imagepng($image, $tempfile);
        // remove image from memory
        imagedestroy($image);
        // get PNG image data
        $retvars = self::_parsepng($tempfile);
        // tidy up by removing temporary image
        unlink($tempfile);
        return $retvars;
    }

    /**
     * Convert the loaded image to a JPEG and then return a structure for the PDF creator.
     * This function requires GD library and write access to the directory defined on K_PATH_CACHE constant.
     * @param $image (image) Image object.
     * @param $quality (int) JPEG quality.
     * @param $tempfile (string) Temporary file name.
     * return image JPEG image object.
     * @public static
     */
    public static function _toJPEG($image, $quality, $tempfile)
    {
        imagejpeg($image, $tempfile, $quality);
        imagedestroy($image);
        $retvars = self::_parsejpeg($tempfile);
        // tidy up by removing temporary image
        unlink($tempfile);
        return $retvars;
    }

    /**
     * Extract info from a JPEG file without using the GD library.
     * @param $file (string) image file to parse
     * @return array structure containing the image data
     * @public static
     */
    public static function _parsejpeg($file)
    {
        // check if is a local file
        if (!@file_exists($file)) {
            // try to encode spaces on filename
            $tfile = str_replace(' ', '%20', $file);
            if (@file_exists($tfile)) {
                $file = $tfile;
            }
        }
        $a = getimagesize($file);
        if (empty($a)) {
            //Missing or incorrect image file
            return false;
        }
        if ($a[2] != 2) {
            // Not a JPEG file
            return false;
        }
        // bits per pixel
        $bpc = isset($a['bits']) ? intval($a['bits']) : 8;
        // number of image channels
        if (!isset($a['channels'])) {
            $channels = 3;
        } else {
            $channels = intval($a['channels']);
        }
        // default colour space
        switch ($channels) {
            case 1:
                $colspace = 'DeviceGray';
                break;

            case 3:
                $colspace = 'DeviceRGB';
                break;

            case 4:
                $colspace = 'DeviceCMYK';
                break;

            default:
                $channels = 3;
                $colspace = 'DeviceRGB';
                break;
        }
        // get file content
        $data = file_get_contents($file);
        // check for embedded ICC profile
        $icc = [];
        $offset = 0;
        while (($pos = strpos($data, "ICC_PROFILE\0", $offset)) !== false) {
            // get ICC sequence length
            $length = (TCPDFSTATIC::_getUSHORT($data, ($pos - 2)) - 16);
            // marker sequence number
            $msn = max(1, ord($data[($pos + 12)]));
            // number of markers (total of APP2 used)
            $nom = max(1, ord($data[($pos + 13)]));
            // get sequence segment
            $icc[($msn - 1)] = substr($data, ($pos + 14), $length);
            // move forward to next sequence
            $offset = ($pos + 14 + $length);
        }
        // order and compact ICC segments
        if (count($icc) > 0) {
            ksort($icc);
            $icc = implode('', $icc);
            if ((ord($icc[36]) != 0x61) or
                (ord($icc[37]) != 0x63) or
                (ord($icc[38]) != 0x73) or
                (ord($icc[39]) != 0x70)
            ) {
                // invalid ICC profile
                $icc = false;
            }
        } else {
            $icc = false;
        }
        return [
            'w' => $a[0],
            'h' => $a[1],
            'ch' => $channels,
            'icc' => $icc,
            'cs' => $colspace,
            'bpc' => $bpc,
            'f' => 'DCTDecode',
            'data' => $data
        ];
    }

    /**
     * Extract info from a PNG file without using the GD library.
     * @param $file (string) image file to parse
     * @return array structure containing the image data
     * @public static
     */
    public static function _parsepng($file)
    {
        $f = @fopen($file, 'rb');
        if ($f === false) {
            // Can't open image file
            return false;
        }
        //Check signature
        if (fread($f, 8) != chr(137).'PNG'.chr(13).chr(10).chr(26).chr(10)) {
            // Not a PNG file
            return false;
        }
        //Read header chunk
        fread($f, 4);
        if (fread($f, 4) != 'IHDR') {
            //Incorrect PNG file
            return false;
        }
        $w = TCPDFSTATIC::_freadint($f);
        $h = TCPDFSTATIC::_freadint($f);
        $bpc = ord(fread($f, 1));
        $ct = ord(fread($f, 1));
        if ($ct == 0) {
            $colspace = 'DeviceGray';
        } elseif ($ct == 2) {
            $colspace = 'DeviceRGB';
        } elseif ($ct == 3) {
            $colspace = 'Indexed';
        } else {
            // alpha channel
            fclose($f);
            return 'pngalpha';
        }
        if (ord(fread($f, 1)) != 0) {
            // Unknown compression method
            fclose($f);
            return false;
        }
        if (ord(fread($f, 1)) != 0) {
            // Unknown filter method
            fclose($f);
            return false;
        }
        if (ord(fread($f, 1)) != 0) {
            // Interlacing not supported
            fclose($f);
            return false;
        }
        fread($f, 4);
        $channels = ($ct == 2 ? 3 : 1);
        $parms = '/DecodeParms << /Predictor 15 /Colors '.$channels.' /BitsPerComponent '.$bpc.' /Columns '.$w.' >>';
        //Scan chunks looking for palette, transparency and image data
        $pal = '';
        $trns = '';
        $data = '';
        $icc = false;
        $n = TCPDFSTATIC::_freadint($f);
        do {
            $type = fread($f, 4);
            if ($type === 'PLTE') {
                // read palette
                $pal = TCPDFSTATIC::rfread($f, $n);
                fread($f, 4);
            } elseif ($type === 'tRNS') {
                // read transparency info
                $t = TCPDFSTATIC::rfread($f, $n);
                if ($ct == 0) { // DeviceGray
                    $trns = [ord($t[1])];
                } elseif ($ct == 2) { // DeviceRGB
                    $trns = [ord($t[1]), ord($t[3]), ord($t[5])];
                } else { // Indexed
                    if ($n > 0) {
                        $trns = [];
                        for ($i = 0; $i < $n; ++ $i) {
                            $trns[] = ord($t{$i});
                        }
                    }
                }
                fread($f, 4);
            } elseif ($type == 'IDAT') {
                // read image data block
                $data .= TCPDFSTATIC::rfread($f, $n);
                fread($f, 4);
            } elseif ($type == 'iCCP') {
                // skip profile name
                $len = 0;
                while ((ord(fread($f, 1)) != 0) and ($len < 80)) {
                    ++$len;
                }
                // get compression method
                if (ord(fread($f, 1)) != 0) {
                    // Unknown filter method
                    fclose($f);
                    return false;
                }
                // read ICC Color Profile
                $icc = TCPDFSTATIC::rfread($f, ($n - $len - 2));
                // decompress profile
                $icc = gzuncompress($icc);
                fread($f, 4);
            } elseif ($type === 'IEND') {
                break;
            } else {
                TCPDFSTATIC::rfread($f, $n + 4);
            }
            $n = TCPDFSTATIC::_freadint($f);
        } while ($n);
        if (($colspace === 'Indexed') and (empty($pal))) {
            // Missing palette
            fclose($f);
            return false;
        }
        fclose($f);
        return [
            'w' => $w,
            'h' => $h,
            'ch' => $channels,
            'icc' => $icc,
            'cs' => $colspace,
            'bpc' => $bpc,
            'f' => 'FlateDecode',
            'parms' => $parms,
            'pal' => $pal,
            'trns' => $trns,
            'data' => $data
        ];
    }
} // END OF TCPDF_IMAGES CLASS

//============================================================+
// END OF FILE
//============================================================+
